sql/stats: TestAutoStatsReadOnlyTables failed

sql/stats.TestAutoStatsReadOnlyTables [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9288589?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9288589?buildTab=artifacts#/) on master @ [143b63a6a27aeb286637dd2e5abddafdf0c51874](https://github.com/cockroachdb/cockroach/commits/143b63a6a27aeb286637dd2e5abddafdf0c51874):


```
=== RUN   TestAutoStatsReadOnlyTables
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/a29a19c212b7b40401087e2c497e19d3/logTestAutoStatsReadOnlyTables3925782756
    test_log_scope.go:79: use -show-logs to present logs inline
    sql_runner.go:103: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/testutils/sqlutils/sql_runner.go:116
        	            				github.com/cockroachdb/cockroach/pkg/testutils/sqlutils/sql_runner.go:311
        	            				github.com/cockroachdb/cockroach/pkg/sql/stats/pkg/sql/stats/automatic_stats_test.go:563
        	Error:      	Received unexpected error:
        	            	query 'SELECT statistics_name, column_names, row_count FROM [SHOW STATISTICS FOR TABLE t.a]': expected:
        	            	(1) attached stack trace
        	            	  -- stack trace:
        	            	  | github.com/cockroachdb/cockroach/pkg/testutils/sqlutils.(*SQLRunner).CheckQueryResultsRetry.func1
        	            	  | 	github.com/cockroachdb/cockroach/pkg/testutils/sqlutils/sql_runner.go:314
        	            	  | github.com/cockroachdb/cockroach/pkg/testutils.SucceedsWithinError.func1
        	            	  | 	github.com/cockroachdb/cockroach/pkg/testutils/soon.go:71
        	            	  | github.com/cockroachdb/cockroach/pkg/util/retry.ForDuration
        	            	  | 	github.com/cockroachdb/cockroach/pkg/util/retry/retry.go:213
        	            	  | github.com/cockroachdb/cockroach/pkg/testutils.SucceedsWithinError
        	            	  | 	github.com/cockroachdb/cockroach/pkg/testutils/soon.go:77
        	            	  | github.com/cockroachdb/cockroach/pkg/testutils/sqlutils.(*SQLRunner).succeedsWithin
        	            	  | 	github.com/cockroachdb/cockroach/pkg/testutils/sqlutils/sql_runner.go:116
        	            	  | github.com/cockroachdb/cockroach/pkg/testutils/sqlutils.(*SQLRunner).CheckQueryResultsRetry
        	            	  | 	github.com/cockroachdb/cockroach/pkg/testutils/sqlutils/sql_runner.go:311
        	            	  | github.com/cockroachdb/cockroach/pkg/sql/stats.TestAutoStatsReadOnlyTables
        	            	  | 	github.com/cockroachdb/cockroach/pkg/sql/stats/pkg/sql/stats/automatic_stats_test.go:563
        	            	  | testing.tRunner
        	            	  | 	GOROOT/src/testing/testing.go:1446
        	            	  | runtime.goexit
        	            	  | 	GOROOT/src/runtime/asm_amd64.s:1594
        	            	Wraps: (2) query 'SELECT statistics_name, column_names, row_count FROM [SHOW STATISTICS FOR TABLE t.a]': expected:
        	            	  | __auto__, {k}, 0
        	            	  | got:
        	            	  | 
        	            	Error types: (1) *withstack.withStack (2) *errutil.leafError
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/a29a19c212b7b40401087e2c497e19d3/logTestAutoStatsReadOnlyTables3925782756
--- FAIL: TestAutoStatsReadOnlyTables (50.45s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-queries
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestAutoStatsReadOnlyTables.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-26086

sql/stats.TestAutoStatsReadOnlyTables [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9312210?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9312210?buildTab=artifacts#/) on master @ [61d1dce768ecf3b37e79f2621d3726d7ff4e95df](https://github.com/cockroachdb/cockroach/commits/61d1dce768ecf3b37e79f2621d3726d7ff4e95df):


```
=== RUN   TestAutoStatsReadOnlyTables
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/a29a19c212b7b40401087e2c497e19d3/logTestAutoStatsReadOnlyTables1151143592
    test_log_scope.go:79: use -show-logs to present logs inline
    sql_runner.go:103: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/testutils/sqlutils/sql_runner.go:116
        	            				github.com/cockroachdb/cockroach/pkg/testutils/sqlutils/sql_runner.go:311
        	            				github.com/cockroachdb/cockroach/pkg/sql/stats/pkg/sql/stats/automatic_stats_test.go:563
        	Error:      	Received unexpected error:
        	            	query 'SELECT statistics_name, column_names, row_count FROM [SHOW STATISTICS FOR TABLE t.a]': expected:
        	            	(1) attached stack trace
        	            	  -- stack trace:
        	            	  | github.com/cockroachdb/cockroach/pkg/testutils/sqlutils.(*SQLRunner).CheckQueryResultsRetry.func1
        	            	  | 	github.com/cockroachdb/cockroach/pkg/testutils/sqlutils/sql_runner.go:314
        	            	  | github.com/cockroachdb/cockroach/pkg/testutils.SucceedsWithinError.func1
        	            	  | 	github.com/cockroachdb/cockroach/pkg/testutils/soon.go:71
        	            	  | github.com/cockroachdb/cockroach/pkg/util/retry.ForDuration
        	            	  | 	github.com/cockroachdb/cockroach/pkg/util/retry/retry.go:213
        	            	  | github.com/cockroachdb/cockroach/pkg/testutils.SucceedsWithinError
        	            	  | 	github.com/cockroachdb/cockroach/pkg/testutils/soon.go:77
        	            	  | github.com/cockroachdb/cockroach/pkg/testutils/sqlutils.(*SQLRunner).succeedsWithin
        	            	  | 	github.com/cockroachdb/cockroach/pkg/testutils/sqlutils/sql_runner.go:116
        	            	  | github.com/cockroachdb/cockroach/pkg/testutils/sqlutils.(*SQLRunner).CheckQueryResultsRetry
        	            	  | 	github.com/cockroachdb/cockroach/pkg/testutils/sqlutils/sql_runner.go:311
        	            	  | github.com/cockroachdb/cockroach/pkg/sql/stats.TestAutoStatsReadOnlyTables
        	            	  | 	github.com/cockroachdb/cockroach/pkg/sql/stats/pkg/sql/stats/automatic_stats_test.go:563
        	            	  | testing.tRunner
        	            	  | 	GOROOT/src/testing/testing.go:1446
        	            	  | runtime.goexit
        	            	  | 	GOROOT/src/runtime/asm_amd64.s:1594
        	            	Wraps: (2) query 'SELECT statistics_name, column_names, row_count FROM [SHOW STATISTICS FOR TABLE t.a]': expected:
        	            	  | __auto__, {k}, 0
        	            	  | got:
        	            	  | 
        	            	Error types: (1) *withstack.withStack (2) *errutil.leafError
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/a29a19c212b7b40401087e2c497e19d3/logTestAutoStatsReadOnlyTables1151143592
--- FAIL: TestAutoStatsReadOnlyTables (50.65s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestAutoStatsReadOnlyTables.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

sql/stats.TestAutoStatsReadOnlyTables [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9359564?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9359564?buildTab=artifacts#/) on master @ [27c7beb75af1a99a862881aca223d3cd29741a65](https://github.com/cockroachdb/cockroach/commits/27c7beb75af1a99a862881aca223d3cd29741a65):


```
=== RUN   TestAutoStatsReadOnlyTables
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/a29a19c212b7b40401087e2c497e19d3/logTestAutoStatsReadOnlyTables597851048
    test_log_scope.go:79: use -show-logs to present logs inline
    sql_runner.go:103: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/testutils/sqlutils/sql_runner.go:116
        	            				github.com/cockroachdb/cockroach/pkg/testutils/sqlutils/sql_runner.go:311
        	            				github.com/cockroachdb/cockroach/pkg/sql/stats/pkg/sql/stats/automatic_stats_test.go:563
        	Error:      	Received unexpected error:
        	            	query 'SELECT statistics_name, column_names, row_count FROM [SHOW STATISTICS FOR TABLE t.a]': expected:
        	            	(1) attached stack trace
        	            	  -- stack trace:
        	            	  | github.com/cockroachdb/cockroach/pkg/testutils/sqlutils.(*SQLRunner).CheckQueryResultsRetry.func1
        	            	  | 	github.com/cockroachdb/cockroach/pkg/testutils/sqlutils/sql_runner.go:314
        	            	  | github.com/cockroachdb/cockroach/pkg/testutils.SucceedsWithinError.func1
        	            	  | 	github.com/cockroachdb/cockroach/pkg/testutils/soon.go:71
        	            	  | github.com/cockroachdb/cockroach/pkg/util/retry.ForDuration
        	            	  | 	github.com/cockroachdb/cockroach/pkg/util/retry/retry.go:213
        	            	  | github.com/cockroachdb/cockroach/pkg/testutils.SucceedsWithinError
        	            	  | 	github.com/cockroachdb/cockroach/pkg/testutils/soon.go:77
        	            	  | github.com/cockroachdb/cockroach/pkg/testutils/sqlutils.(*SQLRunner).succeedsWithin
        	            	  | 	github.com/cockroachdb/cockroach/pkg/testutils/sqlutils/sql_runner.go:116
        	            	  | github.com/cockroachdb/cockroach/pkg/testutils/sqlutils.(*SQLRunner).CheckQueryResultsRetry
        	            	  | 	github.com/cockroachdb/cockroach/pkg/testutils/sqlutils/sql_runner.go:311
        	            	  | github.com/cockroachdb/cockroach/pkg/sql/stats.TestAutoStatsReadOnlyTables
        	            	  | 	github.com/cockroachdb/cockroach/pkg/sql/stats/pkg/sql/stats/automatic_stats_test.go:563
        	            	  | testing.tRunner
        	            	  | 	GOROOT/src/testing/testing.go:1446
        	            	  | runtime.goexit
        	            	  | 	GOROOT/src/runtime/asm_amd64.s:1594
        	            	Wraps: (2) query 'SELECT statistics_name, column_names, row_count FROM [SHOW STATISTICS FOR TABLE t.a]': expected:
        	            	  | __auto__, {k}, 0
        	            	  | got:
        	            	  | 
        	            	Error types: (1) *withstack.withStack (2) *errutil.leafError
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/a29a19c212b7b40401087e2c497e19d3/logTestAutoStatsReadOnlyTables597851048
--- FAIL: TestAutoStatsReadOnlyTables (50.98s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestAutoStatsReadOnlyTables.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

sql/stats.TestAutoStatsReadOnlyTables [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9378639?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9378639?buildTab=artifacts#/) on master @ [99102ddf4b7602788b422366f1acc14b81c64d24](https://github.com/cockroachdb/cockroach/commits/99102ddf4b7602788b422366f1acc14b81c64d24):


```
=== RUN   TestAutoStatsReadOnlyTables
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/a29a19c212b7b40401087e2c497e19d3/logTestAutoStatsReadOnlyTables2834689858
    test_log_scope.go:79: use -show-logs to present logs inline
    sql_runner.go:103: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/testutils/sqlutils/sql_runner.go:116
        	            				github.com/cockroachdb/cockroach/pkg/testutils/sqlutils/sql_runner.go:311
        	            				github.com/cockroachdb/cockroach/pkg/sql/stats/pkg/sql/stats/automatic_stats_test.go:563
        	Error:      	Received unexpected error:
        	            	query 'SELECT statistics_name, column_names, row_count FROM [SHOW STATISTICS FOR TABLE t.a]': expected:
        	            	(1) attached stack trace
        	            	  -- stack trace:
        	            	  | github.com/cockroachdb/cockroach/pkg/testutils/sqlutils.(*SQLRunner).CheckQueryResultsRetry.func1
        	            	  | 	github.com/cockroachdb/cockroach/pkg/testutils/sqlutils/sql_runner.go:314
        	            	  | github.com/cockroachdb/cockroach/pkg/testutils.SucceedsWithinError.func1
        	            	  | 	github.com/cockroachdb/cockroach/pkg/testutils/soon.go:71
        	            	  | github.com/cockroachdb/cockroach/pkg/util/retry.ForDuration
        	            	  | 	github.com/cockroachdb/cockroach/pkg/util/retry/retry.go:213
        	            	  | github.com/cockroachdb/cockroach/pkg/testutils.SucceedsWithinError
        	            	  | 	github.com/cockroachdb/cockroach/pkg/testutils/soon.go:77
        	            	  | github.com/cockroachdb/cockroach/pkg/testutils/sqlutils.(*SQLRunner).succeedsWithin
        	            	  | 	github.com/cockroachdb/cockroach/pkg/testutils/sqlutils/sql_runner.go:116
        	            	  | github.com/cockroachdb/cockroach/pkg/testutils/sqlutils.(*SQLRunner).CheckQueryResultsRetry
        	            	  | 	github.com/cockroachdb/cockroach/pkg/testutils/sqlutils/sql_runner.go:311
        	            	  | github.com/cockroachdb/cockroach/pkg/sql/stats.TestAutoStatsReadOnlyTables
        	            	  | 	github.com/cockroachdb/cockroach/pkg/sql/stats/pkg/sql/stats/automatic_stats_test.go:563
        	            	  | testing.tRunner
        	            	  | 	GOROOT/src/testing/testing.go:1446
        	            	  | runtime.goexit
        	            	  | 	GOROOT/src/runtime/asm_amd64.s:1594
        	            	Wraps: (2) query 'SELECT statistics_name, column_names, row_count FROM [SHOW STATISTICS FOR TABLE t.a]': expected:
        	            	  | __auto__, {k}, 0
        	            	  | got:
        	            	  | 
        	            	Error types: (1) *withstack.withStack (2) *errutil.leafError
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/a29a19c212b7b40401087e2c497e19d3/logTestAutoStatsReadOnlyTables2834689858
--- FAIL: TestAutoStatsReadOnlyTables (51.65s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestAutoStatsReadOnlyTables.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

sql/stats.TestAutoStatsReadOnlyTables [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9388453?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9388453?buildTab=artifacts#/) on master @ [1e27c65fb5a3810a5e2285cf07b45d95d401cca6](https://github.com/cockroachdb/cockroach/commits/1e27c65fb5a3810a5e2285cf07b45d95d401cca6):


```
=== RUN   TestAutoStatsReadOnlyTables
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/a29a19c212b7b40401087e2c497e19d3/logTestAutoStatsReadOnlyTables464572653
    test_log_scope.go:79: use -show-logs to present logs inline
    sql_runner.go:103: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/testutils/sqlutils/sql_runner.go:116
        	            				github.com/cockroachdb/cockroach/pkg/testutils/sqlutils/sql_runner.go:311
        	            				github.com/cockroachdb/cockroach/pkg/sql/stats/pkg/sql/stats/automatic_stats_test.go:563
        	Error:      	Received unexpected error:
        	            	query 'SELECT statistics_name, column_names, row_count FROM [SHOW STATISTICS FOR TABLE t.a]': expected:
        	            	(1) attached stack trace
        	            	  -- stack trace:
        	            	  | github.com/cockroachdb/cockroach/pkg/testutils/sqlutils.(*SQLRunner).CheckQueryResultsRetry.func1
        	            	  | 	github.com/cockroachdb/cockroach/pkg/testutils/sqlutils/sql_runner.go:314
        	            	  | github.com/cockroachdb/cockroach/pkg/testutils.SucceedsWithinError.func1
        	            	  | 	github.com/cockroachdb/cockroach/pkg/testutils/soon.go:71
        	            	  | github.com/cockroachdb/cockroach/pkg/util/retry.ForDuration
        	            	  | 	github.com/cockroachdb/cockroach/pkg/util/retry/retry.go:213
        	            	  | github.com/cockroachdb/cockroach/pkg/testutils.SucceedsWithinError
        	            	  | 	github.com/cockroachdb/cockroach/pkg/testutils/soon.go:77
        	            	  | github.com/cockroachdb/cockroach/pkg/testutils/sqlutils.(*SQLRunner).succeedsWithin
        	            	  | 	github.com/cockroachdb/cockroach/pkg/testutils/sqlutils/sql_runner.go:116
        	            	  | github.com/cockroachdb/cockroach/pkg/testutils/sqlutils.(*SQLRunner).CheckQueryResultsRetry
        	            	  | 	github.com/cockroachdb/cockroach/pkg/testutils/sqlutils/sql_runner.go:311
        	            	  | github.com/cockroachdb/cockroach/pkg/sql/stats.TestAutoStatsReadOnlyTables
        	            	  | 	github.com/cockroachdb/cockroach/pkg/sql/stats/pkg/sql/stats/automatic_stats_test.go:563
        	            	  | testing.tRunner
        	            	  | 	GOROOT/src/testing/testing.go:1446
        	            	  | runtime.goexit
        	            	  | 	GOROOT/src/runtime/asm_amd64.s:1594
        	            	Wraps: (2) query 'SELECT statistics_name, column_names, row_count FROM [SHOW STATISTICS FOR TABLE t.a]': expected:
        	            	  | __auto__, {k}, 0
        	            	  | got:
        	            	  | 
        	            	Error types: (1) *withstack.withStack (2) *errutil.leafError
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/a29a19c212b7b40401087e2c497e19d3/logTestAutoStatsReadOnlyTables464572653
--- FAIL: TestAutoStatsReadOnlyTables (51.22s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #100398 sql/stats: TestAutoStatsReadOnlyTables failed [C-test-failure O-robot T-sql-queries branch-release-23.1]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestAutoStatsReadOnlyTables.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

sql/stats.TestAutoStatsReadOnlyTables [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9399286?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9399286?buildTab=artifacts#/) on master @ [66dfb23ae120c688df7652ce9d8a95215735fc0d](https://github.com/cockroachdb/cockroach/commits/66dfb23ae120c688df7652ce9d8a95215735fc0d):


```
=== RUN   TestAutoStatsReadOnlyTables
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/a29a19c212b7b40401087e2c497e19d3/logTestAutoStatsReadOnlyTables3217469205
    test_log_scope.go:79: use -show-logs to present logs inline
    sql_runner.go:103: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/testutils/sqlutils/sql_runner.go:116
        	            				github.com/cockroachdb/cockroach/pkg/testutils/sqlutils/sql_runner.go:311
        	            				github.com/cockroachdb/cockroach/pkg/sql/stats/pkg/sql/stats/automatic_stats_test.go:563
        	Error:      	Received unexpected error:
        	            	query 'SELECT statistics_name, column_names, row_count FROM [SHOW STATISTICS FOR TABLE t.a]': expected:
        	            	(1) attached stack trace
        	            	  -- stack trace:
        	            	  | github.com/cockroachdb/cockroach/pkg/testutils/sqlutils.(*SQLRunner).CheckQueryResultsRetry.func1
        	            	  | 	github.com/cockroachdb/cockroach/pkg/testutils/sqlutils/sql_runner.go:314
        	            	  | github.com/cockroachdb/cockroach/pkg/testutils.SucceedsWithinError.func1
        	            	  | 	github.com/cockroachdb/cockroach/pkg/testutils/soon.go:71
        	            	  | github.com/cockroachdb/cockroach/pkg/util/retry.ForDuration
        	            	  | 	github.com/cockroachdb/cockroach/pkg/util/retry/retry.go:213
        	            	  | github.com/cockroachdb/cockroach/pkg/testutils.SucceedsWithinError
        	            	  | 	github.com/cockroachdb/cockroach/pkg/testutils/soon.go:77
        	            	  | github.com/cockroachdb/cockroach/pkg/testutils/sqlutils.(*SQLRunner).succeedsWithin
        	            	  | 	github.com/cockroachdb/cockroach/pkg/testutils/sqlutils/sql_runner.go:116
        	            	  | github.com/cockroachdb/cockroach/pkg/testutils/sqlutils.(*SQLRunner).CheckQueryResultsRetry
        	            	  | 	github.com/cockroachdb/cockroach/pkg/testutils/sqlutils/sql_runner.go:311
        	            	  | github.com/cockroachdb/cockroach/pkg/sql/stats.TestAutoStatsReadOnlyTables
        	            	  | 	github.com/cockroachdb/cockroach/pkg/sql/stats/pkg/sql/stats/automatic_stats_test.go:563
        	            	  | testing.tRunner
        	            	  | 	GOROOT/src/testing/testing.go:1446
        	            	  | runtime.goexit
        	            	  | 	GOROOT/src/runtime/asm_amd64.s:1594
        	            	Wraps: (2) query 'SELECT statistics_name, column_names, row_count FROM [SHOW STATISTICS FOR TABLE t.a]': expected:
        	            	  | __auto__, {k}, 0
        	            	  | got:
        	            	  | 
        	            	Error types: (1) *withstack.withStack (2) *errutil.leafError
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/a29a19c212b7b40401087e2c497e19d3/logTestAutoStatsReadOnlyTables3217469205
--- FAIL: TestAutoStatsReadOnlyTables (52.33s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #100398 sql/stats: TestAutoStatsReadOnlyTables failed [C-test-failure O-robot T-sql-queries branch-release-23.1]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestAutoStatsReadOnlyTables.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

sql/stats.TestAutoStatsReadOnlyTables [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9422917?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9422917?buildTab=artifacts#/) on master @ [07c7d4b3abfa1dc79d8fda385132a27c81665fee](https://github.com/cockroachdb/cockroach/commits/07c7d4b3abfa1dc79d8fda385132a27c81665fee):


```
=== RUN   TestAutoStatsReadOnlyTables
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/a29a19c212b7b40401087e2c497e19d3/logTestAutoStatsReadOnlyTables1507517540
    test_log_scope.go:79: use -show-logs to present logs inline
    sql_runner.go:103: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/testutils/sqlutils/sql_runner.go:116
        	            				github.com/cockroachdb/cockroach/pkg/testutils/sqlutils/sql_runner.go:311
        	            				github.com/cockroachdb/cockroach/pkg/sql/stats/pkg/sql/stats/automatic_stats_test.go:563
        	Error:      	Received unexpected error:
        	            	query 'SELECT statistics_name, column_names, row_count FROM [SHOW STATISTICS FOR TABLE t.a]': expected:
        	            	(1) attached stack trace
        	            	  -- stack trace:
        	            	  | github.com/cockroachdb/cockroach/pkg/testutils/sqlutils.(*SQLRunner).CheckQueryResultsRetry.func1
        	            	  | 	github.com/cockroachdb/cockroach/pkg/testutils/sqlutils/sql_runner.go:314
        	            	  | github.com/cockroachdb/cockroach/pkg/testutils.SucceedsWithinError.func1
        	            	  | 	github.com/cockroachdb/cockroach/pkg/testutils/soon.go:71
        	            	  | github.com/cockroachdb/cockroach/pkg/util/retry.ForDuration
        	            	  | 	github.com/cockroachdb/cockroach/pkg/util/retry/retry.go:213
        	            	  | github.com/cockroachdb/cockroach/pkg/testutils.SucceedsWithinError
        	            	  | 	github.com/cockroachdb/cockroach/pkg/testutils/soon.go:77
        	            	  | github.com/cockroachdb/cockroach/pkg/testutils/sqlutils.(*SQLRunner).succeedsWithin
        	            	  | 	github.com/cockroachdb/cockroach/pkg/testutils/sqlutils/sql_runner.go:116
        	            	  | github.com/cockroachdb/cockroach/pkg/testutils/sqlutils.(*SQLRunner).CheckQueryResultsRetry
        	            	  | 	github.com/cockroachdb/cockroach/pkg/testutils/sqlutils/sql_runner.go:311
        	            	  | github.com/cockroachdb/cockroach/pkg/sql/stats.TestAutoStatsReadOnlyTables
        	            	  | 	github.com/cockroachdb/cockroach/pkg/sql/stats/pkg/sql/stats/automatic_stats_test.go:563
        	            	  | testing.tRunner
        	            	  | 	GOROOT/src/testing/testing.go:1446
        	            	  | runtime.goexit
        	            	  | 	GOROOT/src/runtime/asm_amd64.s:1594
        	            	Wraps: (2) query 'SELECT statistics_name, column_names, row_count FROM [SHOW STATISTICS FOR TABLE t.a]': expected:
        	            	  | __auto__, {k}, 0
        	            	  | got:
        	            	  | 
        	            	Error types: (1) *withstack.withStack (2) *errutil.leafError
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/a29a19c212b7b40401087e2c497e19d3/logTestAutoStatsReadOnlyTables1507517540
--- FAIL: TestAutoStatsReadOnlyTables (55.96s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #100576 sql/stats: TestAutoStatsReadOnlyTables failed [C-test-failure O-robot T-sql-queries branch-release-23.1]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestAutoStatsReadOnlyTables.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

sql/stats.TestAutoStatsReadOnlyTables [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9467422?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9467422?buildTab=artifacts#/) on master @ [530100fd39cc722bc324bfb3869a325622258fb3](https://github.com/cockroachdb/cockroach/commits/530100fd39cc722bc324bfb3869a325622258fb3):


```
=== RUN   TestAutoStatsReadOnlyTables
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/a29a19c212b7b40401087e2c497e19d3/logTestAutoStatsReadOnlyTables1867691834
    test_log_scope.go:79: use -show-logs to present logs inline
    sql_runner.go:103: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/testutils/sqlutils/sql_runner.go:116
        	            				github.com/cockroachdb/cockroach/pkg/testutils/sqlutils/sql_runner.go:311
        	            				github.com/cockroachdb/cockroach/pkg/sql/stats/pkg/sql/stats/automatic_stats_test.go:563
        	Error:      	Received unexpected error:
        	            	query 'SELECT statistics_name, column_names, row_count FROM [SHOW STATISTICS FOR TABLE t.a]': expected:
        	            	(1) attached stack trace
        	            	  -- stack trace:
        	            	  | github.com/cockroachdb/cockroach/pkg/testutils/sqlutils.(*SQLRunner).CheckQueryResultsRetry.func1
        	            	  | 	github.com/cockroachdb/cockroach/pkg/testutils/sqlutils/sql_runner.go:314
        	            	  | github.com/cockroachdb/cockroach/pkg/testutils.SucceedsWithinError.func1
        	            	  | 	github.com/cockroachdb/cockroach/pkg/testutils/soon.go:71
        	            	  | github.com/cockroachdb/cockroach/pkg/util/retry.ForDuration
        	            	  | 	github.com/cockroachdb/cockroach/pkg/util/retry/retry.go:213
        	            	  | github.com/cockroachdb/cockroach/pkg/testutils.SucceedsWithinError
        	            	  | 	github.com/cockroachdb/cockroach/pkg/testutils/soon.go:77
        	            	  | github.com/cockroachdb/cockroach/pkg/testutils/sqlutils.(*SQLRunner).succeedsWithin
        	            	  | 	github.com/cockroachdb/cockroach/pkg/testutils/sqlutils/sql_runner.go:116
        	            	  | github.com/cockroachdb/cockroach/pkg/testutils/sqlutils.(*SQLRunner).CheckQueryResultsRetry
        	            	  | 	github.com/cockroachdb/cockroach/pkg/testutils/sqlutils/sql_runner.go:311
        	            	  | github.com/cockroachdb/cockroach/pkg/sql/stats.TestAutoStatsReadOnlyTables
        	            	  | 	github.com/cockroachdb/cockroach/pkg/sql/stats/pkg/sql/stats/automatic_stats_test.go:563
        	            	  | testing.tRunner
        	            	  | 	GOROOT/src/testing/testing.go:1446
        	            	  | runtime.goexit
        	            	  | 	GOROOT/src/runtime/asm_amd64.s:1594
        	            	Wraps: (2) query 'SELECT statistics_name, column_names, row_count FROM [SHOW STATISTICS FOR TABLE t.a]': expected:
        	            	  | __auto__, {k}, 0
        	            	  | got:
        	            	  | 
        	            	Error types: (1) *withstack.withStack (2) *errutil.leafError
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/a29a19c212b7b40401087e2c497e19d3/logTestAutoStatsReadOnlyTables1867691834
--- FAIL: TestAutoStatsReadOnlyTables (52.58s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestAutoStatsReadOnlyTables.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/sql/importer/import_stmt_test.go
+++ b/pkg/sql/importer/import_stmt_test.go
@@ -6598,7 +6598,19 @@ func TestCreateStatsAfterImport(t *testing.T) {
 	const nodes = 1
 	ctx := context.Background()
 	baseDir := datapathutils.TestDataPath(t)
-	args := base.TestServerArgs{ExternalIODir: baseDir}
+
+	// Disable stats collection on system tables before the cluster is started,
+	// otherwise there is a race condition where stats may be collected before we
+	// can disable them with `SET CLUSTER SETTING`. We disable stats collection on
+	// system tables so that we can collect statistics on the imported tables
+	// within the retry time limit.
+	st := cluster.MakeClusterSettings()
+	stats.AutomaticStatisticsOnSystemTables.Override(context.Background(), &st.SV, false)
+	args := base.TestServerArgs{
+		Settings:          st,
+		DefaultTestTenant: base.TestTenantDisabled,
+		ExternalIODir:     baseDir,
+	}
 	tc := serverutils.StartNewTestCluster(t, nodes, base.TestClusterArgs{ServerArgs: args})
 	defer tc.Stopper().Stop(ctx)
 	conn := tc.ServerConn(0)
@@ -6609,20 +6621,18 @@ func TestCreateStatsAfterImport(t *testing.T) {
 	sqlDB.Exec(t, "IMPORT PGDUMP ($1) WITH ignore_unsupported_statements", "nodelocal://1/cockroachdump/dump.sql")
 
 	// Verify that statistics have been created.
-	// Depending on timing, the statistics name may either be __auto__ or
-	// __import__, so we don't check the name in the results.
 	sqlDB.CheckQueryResultsRetry(t,
-		`SELECT column_names, row_count, distinct_count, null_count
+		`SELECT statistics_name, column_names, row_count, distinct_count, null_count
 	  FROM [SHOW STATISTICS FOR TABLE t]`,
 		[][]string{
-			{"{i}", "2", "2", "0"},
-			{"{t}", "2", "2", "0"},
+			{"__auto__", "{i}", "2", "2", "0"},
+			{"__auto__", "{t}", "2", "2", "0"},
 		})
 	sqlDB.CheckQueryResultsRetry(t,
-		`SELECT column_names, row_count, distinct_count, null_count
+		`SELECT statistics_name, column_names, row_count, distinct_count, null_count
 	  FROM [SHOW STATISTICS FOR TABLE a]`,
 		[][]string{
-			{"{i}", "1", "1", "0"},
+			{"__auto__", "{i}", "1", "1", "0"},
 		})
 }
 

--- a/pkg/sql/stats/automatic_stats.go
+++ b/pkg/sql/stats/automatic_stats.go
@@ -573,6 +573,7 @@ WHERE
 	AND (
 			crdb_internal.pb_to_json('cockroach.sql.sqlbase.Descriptor', d.descriptor, false)->'table'->>'viewQuery'
 		) IS NULL
+	%s
 	%s`
 
 	explicitlyEnabledTablesPredicate = `AND
@@ -586,6 +587,10 @@ WHERE
 	 OR crdb_internal.pb_to_json('cockroach.sql.sqlbase.Descriptor',
 		d.descriptor, false)->'table'->'autoStatsSettings' ->> 'enabled' = 'true'
 	)`
+
+	autoStatsOnSystemTablesEnabledPredicate = `AND TRUE`
+
+	autoStatsOnSystemTablesDisabledPredicate = `AND tbl.database_name != 'system'`
 )
 
 func (r *Refresher) getApplicableTables(
@@ -631,29 +636,24 @@ func (r *Refresher) getApplicableTables(
 func (r *Refresher) ensureAllTables(
 	ctx context.Context, settings *settings.Values, initialTableCollectionDelay time.Duration,
 ) {
+	// A table-level setting of sql_stats_automatic_collection_enabled of null,
+	// meaning not set, or true qualifies rows we're interested in.
+	autoStatsPredicate := autoStatsEnabledOrNotSpecifiedPredicate
 	if !AutomaticStatisticsClusterMode.Get(settings) {
-		// Use a historical read so as to disable txn contention resolution.
-		// A table-level setting of sql_stats_automatic_collection_enabled=true is
-		// checked and only those tables are included in this scan.
-		getTablesWithAutoStatsExplicitlyEnabledQuery := fmt.Sprintf(
-			getAllTablesTemplateSQL,
-			initialTableCollectionDelay,
-			keys.TableStatisticsTableID, keys.LeaseTableID, keys.JobsTableID, keys.ScheduledJobsTableID,
-			explicitlyEnabledTablesPredicate,
-		)
-		r.getApplicableTables(ctx, getTablesWithAutoStatsExplicitlyEnabledQuery,
-			"get-tables-with-autostats-explicitly-enabled", false)
-		return
+		autoStatsPredicate = explicitlyEnabledTablesPredicate
+	}
+
+	systemTablesPredicate := autoStatsOnSystemTablesEnabledPredicate
+	if !AutomaticStatisticsOnSystemTables.Get(settings) {
+		systemTablesPredicate = autoStatsOnSystemTablesDisabledPredicate
 	}
 
 	// Use a historical read so as to disable txn contention resolution.
-	// A table-level setting of sql_stats_automatic_collection_enabled of null,
-	// meaning not set, or true qualifies rows we're interested in.
 	getAllTablesQuery := fmt.Sprintf(
 		getAllTablesTemplateSQL,
 		initialTableCollectionDelay,
 		keys.TableStatisticsTableID, keys.LeaseTableID, keys.JobsTableID, keys.ScheduledJobsTableID,
-		autoStatsEnabledOrNotSpecifiedPredicate,
+		autoStatsPredicate, systemTablesPredicate,
 	)
 	r.getApplicableTables(ctx, getAllTablesQuery,
 		"get-tables", false)
@@ -823,17 +823,19 @@ func (r *Refresher) maybeRefreshStats(
 
 func (r *Refresher) refreshStats(ctx context.Context, tableID descpb.ID, asOf time.Duration) error {
 	// Create statistics for all default column sets on the given table.
+	stmt := fmt.Sprintf(
+		"CREATE STATISTICS %s FROM [%d] WITH OPTIONS THROTTLING %g AS OF SYSTEM TIME '-%s'",
+		jobspb.AutoStatsName,
+		tableID,
+		AutomaticStatisticsMaxIdleTime.Get(&r.st.SV),
+		asOf.String(),
+	)
+	log.Infof(ctx, "automatically executing %q", stmt)
 	_ /* rows */, err := r.ex.Exec(
 		ctx,
 		"create-stats",
 		nil, /* txn */
-		fmt.Sprintf(
-			"CREATE STATISTICS %s FROM [%d] WITH OPTIONS THROTTLING %g AS OF SYSTEM TIME '-%s'",
-			jobspb.AutoStatsName,
-			tableID,
-			AutomaticStatisticsMaxIdleTime.Get(&r.st.SV),
-			asOf.String(),
-		),
+		stmt,
 	)
 	return err
 }

--- a/pkg/sql/stats/automatic_stats_test.go
+++ b/pkg/sql/stats/automatic_stats_test.go
@@ -564,6 +564,7 @@ func TestAutoStatsReadOnlyTables(t *testing.T) {
 
 	st := cluster.MakeTestingClusterSettings()
 	AutomaticStatisticsClusterMode.Override(ctx, &st.SV, false)
+	AutomaticStatisticsOnSystemTables.Override(ctx, &st.SV, false)
 	evalCtx := eval.NewTestingEvalContext(st)
 	defer evalCtx.Stop(ctx)
 
@@ -707,6 +708,7 @@ func TestMutationsAndSettingOverrideChannels(t *testing.T) {
 	evalCtx := eval.NewTestingEvalContext(st)
 	defer evalCtx.Stop(ctx)
 
+	AutomaticStatisticsOnSystemTables.Override(ctx, &st.SV, false)
 	AutomaticStatisticsClusterMode.Override(ctx, &st.SV, true)
 	r := Refresher{
 		st:        st,

--- a/pkg/sql/stats/automatic_stats_test.go
+++ b/pkg/sql/stats/automatic_stats_test.go
@@ -206,33 +206,64 @@ func TestEnsureAllTablesQueries(t *testing.T) {
 	numUserTablesWithStats := 2
 
 	// This now includes 36 system tables as well as the 2 created above.
-	if err := checkAllTablesCount(ctx, systemTablesWithStats+numUserTablesWithStats, r); err != nil {
+	if err := checkAllTablesCount(
+		ctx, true /* systemTables */, systemTablesWithStats+numUserTablesWithStats, r,
+	); err != nil {
+		t.Fatal(err)
+	}
+	if err := checkAllTablesCount(
+		ctx, false /* systemTables */, numUserTablesWithStats, r,
+	); err != nil {
 		t.Fatal(err)
 	}
 	if err := checkExplicitlyEnabledTablesCount(ctx, 0, r); err != nil {
 		t.Fatal(err)
 	}
+
 	sqlRun.Exec(t,
 		`ALTER TABLE t.a SET (sql_stats_automatic_collection_enabled = true)`)
-	if err := checkAllTablesCount(ctx, systemTablesWithStats+numUserTablesWithStats, r); err != nil {
+	if err := checkAllTablesCount(
+		ctx, true /* systemTables */, systemTablesWithStats+numUserTablesWithStats, r,
+	); err != nil {
+		t.Fatal(err)
+	}
+	if err := checkAllTablesCount(
+		ctx, false /* systemTables */, numUserTablesWithStats, r,
+	); err != nil {
 		t.Fatal(err)
 	}
 	if err := checkExplicitlyEnabledTablesCount(ctx, 1, r); err != nil {
 		t.Fatal(err)
 	}
+
 	sqlRun.Exec(t,
 		`ALTER TABLE t.b SET (sql_stats_automatic_collection_enabled = false)`)
 	numUserTablesWithStats--
-	if err := checkAllTablesCount(ctx, systemTablesWithStats+numUserTablesWithStats, r); err != nil {
+	if err := checkAllTablesCount(
+		ctx, true /* systemTables */, systemTablesWithStats+numUserTablesWithStats, r,
+	); err != nil {
+		t.Fatal(err)
+	}
+	if err := checkAllTablesCount(
+		ctx, false /* systemTables */, numUserTablesWithStats, r,
+	); err != nil {
 		t.Fatal(err)
 	}
 	if err := checkExplicitlyEnabledTablesCount(ctx, 1, r); err != nil {
 		t.Fatal(err)
 	}
+
 	sqlRun.Exec(t,
 		`ALTER TABLE t.a SET (sql_stats_automatic_collection_enabled = false)`)
 	numUserTablesWithStats--
-	if err := checkAllTablesCount(ctx, systemTablesWithStats+numUserTablesWithStats, r); err != nil {
+	if err := checkAllTablesCount(
+		ctx, true /* systemTables */, systemTablesWithStats+numUserTablesWithStats, r,
+	); err != nil {
+		t.Fatal(err)
+	}
+	if err := checkAllTablesCount(
+		ctx, false /* systemTables */, numUserTablesWithStats, r,
+	); err != nil {
 		t.Fatal(err)
 	}
 	if err := checkExplicitlyEnabledTablesCount(ctx, numUserTablesWithStats, r); err != nil {
@@ -240,13 +271,17 @@ func TestEnsureAllTablesQueries(t *testing.T) {
 	}
 }
 
-func checkAllTablesCount(ctx context.Context, expected int, r *Refresher) error {
+func checkAllTablesCount(ctx context.Context, systemTables bool, expected int, r *Refresher) error {
 	const collectionDelay = time.Microsecond
+	systemTablesPredicate := autoStatsOnSystemTablesEnabledPredicate
+	if !systemTables {
+		systemTablesPredicate = autoStatsOnSystemTablesDisabledPredicate
+	}
 	getAllTablesQuery := fmt.Sprintf(
 		getAllTablesTemplateSQL,
 		collectionDelay,
 		keys.TableStatisticsTableID, keys.LeaseTableID, keys.JobsTableID, keys.ScheduledJobsTableID,
-		autoStatsEnabledOrNotSpecifiedPredicate,
+		autoStatsEnabledOrNotSpecifiedPredicate, systemTablesPredicate,
 	)
 	r.getApplicableTables(ctx, getAllTablesQuery,
 		"get-tables", true)
@@ -263,7 +298,7 @@ func checkExplicitlyEnabledTablesCount(ctx context.Context, expected int, r *Ref
 		getAllTablesTemplateSQL,
 		collectionDelay,
 		keys.TableStatisticsTableID, keys.LeaseTableID, keys.JobsTableID, keys.ScheduledJobsTableID,
-		explicitlyEnabledTablesPredicate,
+		explicitlyEnabledTablesPredicate, autoStatsOnSystemTablesEnabledPredicate,
 	)
 	r.getApplicableTables(ctx, getTablesWithAutoStatsExplicitlyEnabledQuery,
 		"get-tables-with-autostats-explicitly-enabled", true)

